subprojects/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -64,10 +64,14 @@ public void removeListener(Object listener) {
             ListenerDetails details = allListeners.remove(listener);
             if (details != null) {
                 details.disconnect();
-                for (EventBroadcast<?> broadcaster : broadcasters.values()) {
-                    broadcaster.maybeRemove(details);
+                details.listenerLock.lock();
```

Comment:
You are right, I have a test which reproduces this case. Let me see if I can fix it. Good catch!